src/Symfony/Component/Security/Core/Encoder/Argon2iPasswordEncoder.php

```
@@ -0,0 +1,96 @@
+            return $this->encodePasswordNative($raw);
+        }
+        if (function_exists('sodium_crypto_pwhash_str')) {
+            return $this->encodePasswordSodiumFunction($raw);
+        }
+        if (extension_loaded('libsodium')) {
+            return $this->encodePasswordSodiumExtension($raw);
+        }
+
+        throw new \LogicException(
+            'Argon2i algorithm is not supported. Please install libsodium extension or upgrade to PHP 7.2+.'
+        );
+    }
+
+    private function encodePasswordNative($raw)
+    {
+        $hash = password_hash($raw, PASSWORD_ARGON2I);
+
+        return $hash;
+    }
+
+    private function encodePasswordSodiumFunction($raw)
+    {
+        $hash = \sodium_crypto_pwhash_str(
+            $raw,
+            SODIUM_CRYPTO_PWHASH_OPSLIMIT_INTERACTIVE,
+            SODIUM_CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE
+        );
+        \sodium_memzero($raw);
+
+        return $hash;
+    }
+
+    private function encodePasswordSodiumExtension($raw)
+    {
+        $hash = \Sodium\crypto_pwhash_str(
+            $raw,
+            \Sodium\CRYPTO_PWHASH_OPSLIMIT_INTERACTIVE,
+            \Sodium\CRYPTO_PWHASH_MEMLIMIT_INTERACTIVE
+        );
+        \Sodium\memzero($raw);
+
+        return $hash;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function isPasswordValid($encoded, $raw, $salt)
+    {
+        $valid = false;
+        if (PHP_VERSION_ID >= 70200 && defined('PASSWORD_ARGON2I')) {
+            $valid = !$this->isPasswordTooLong($raw) && password_verify($raw, $encoded);
+        } elseif (function_exists('sodium_crypto_pwhash_str_verify')) {
+            $valid = !$this->isPasswordTooLong($raw) && \sodium_crypto_pwhash_str_verify($encoded, $raw);
+            \sodium_memzero($raw);
+        } elseif (extension_loaded('libsodium')) {
+            $valid = !$this->isPasswordTooLong($raw) && \Sodium\crypto_pwhash_str_verify($encoded, $raw);
+            \Sodium\memzero($raw);
+        }
```

If there are no way to check if password is valid, we should throw an exception instead of just returning `false`. At least, that's what other encoder do as this makes debugging much easier.